### ISSUE TITLE
distributor might start serving writes with an empty HA tracker state

### ISSUE THREAD

**Issue by dimitarvdimitrov** [grafana/mimir#5796]
#### Describe the bug
 The distributor doesn't wait for the [HA tracker](https://grafana.com/docs/mimir/latest/configure/configure-high-availability-deduplication/) to complete initializing before starting to serve writes. This means that the distributor can start serving writes with an empty HA tracker state. An empty state will lead to accepting any sample from any HA replica.

#### Details

The HA tracker watches the KV store for changes in case any other distributor replica updates the state.

https://github.com/grafana/mimir/blob/9af647e73b2e75cdffc054770850317dac826697/pkg/distributor/ha_tracker.go#L230

The HA tracker also syncs its own state to the KV store periodically, which would trigger a WatchKey event and cause the tracker to update its tracker.

https://github.com/grafana/mimir/blob/9af647e73b2e75cdffc054770850317dac826697/pkg/distributor/ha_tracker.go#L279-L283

However, the first update of the KV store might happen as late as after one `UpdateTimeout`, which is 15 seconds by default. If there are only a few other distributors, then the KV store changes might come near the end of the 15 seconds, which can be enough for the distributor to start serving writes and accept a write from a non-primary replica.

#### How to fix

The distributor can list all clusters' keys upon startup and update its internal cache, similar to how it's done for each individual replica/cluster [here](https://github.com/grafana/mimir/blob/9af647e73b2e75cdffc054770850317dac826697/pkg/distributor/ha_tracker.go#L231-L263). This can be done as the `StartingFn` of the HA tracker service [here](https://github.com/grafana/mimir/blob/9af647e73b2e75cdffc054770850317dac826697/pkg/distributor/ha_tracker.go#L207) so the dependants of the HA tracker don't start before it has synced the state from the KV store.

**Comment by NickAnge** [comment]
Hello @dimitarvdimitrov. I am planning to work on this issue as part of https://github.com/grafana/pir-actions/issues/94. So, I will investigate/implement the cache population in the context of memberlist kv store.  I will assign it to me, but please let me know if you want me to create a separate issue
